packages/wallet/src/services/WalletAssetProvider.ts

```diff
@@ -0,0 +1,214 @@
+import { Logger } from 'ts-log';
+import { Observable, firstValueFrom } from 'rxjs';
+import { isNotNil } from '@cardano-sdk/util';
+
+export interface AssetProviderContext {
+  assetProvider: AssetProvider;
+  assetInfo$: Observable<Assets>;
+  tx: Cardano.Tx;
+  logger: Logger;
+}
+
+const tryCip68Metadata = (
+  policyId: Cardano.PolicyId,
+  name: Cardano.AssetName,
+  tx: Cardano.Tx,
+  logger: Logger
+): Asset.NftMetadata | null => {
+  const decoded = Asset.AssetNameLabel.decode(name);
+  if (decoded?.label === Asset.AssetNameLabelNum.UserNFT) {
+    const referenceAssetId = Cardano.AssetId.fromParts(
+      policyId,
+      Asset.AssetNameLabel.encode(decoded.content, Asset.AssetNameLabelNum.ReferenceNFT)
+    );
+
+    // TODO: It is possible that the reference NFT is not in one of the outputs of the transaction and was previously minted. We
+    // need a way to find the reference NFT TxOut from the current active UTXO set on the network.
+    for (const output of tx.body.outputs) {
+      if (output.value.assets?.get(referenceAssetId)) {
+        return Asset.NftMetadata.fromPlutusData(output.datum, logger);
+      }
+    }
+  }
+
+  return null;
+};
+
+const getNftMetadata = (
+  name: Cardano.AssetName,
+  policyId: Cardano.PolicyId,
+  tx: Cardano.Tx,
+  logger: Logger
+): Asset.NftMetadata | null => {
+  // First, try CIP-25
+  let metadata = tx.auxiliaryData?.blob
+    ? Asset.NftMetadata.fromMetadatum({ name, policyId }, tx.auxiliaryData.blob, logger)
+    : undefined;
+
+  // If metadata is not found, try CIP-68
+  if (!metadata) {
+    metadata = tryCip68Metadata(policyId, name, tx, logger);
+  }
+
+  return metadata;
+};
+
+const createAssetInfo = (assetId: Cardano.AssetId, amount: bigint, tx: Cardano.Tx, logger: Logger): Asset.AssetInfo => {
+  const name = Cardano.AssetId.getAssetName(assetId);
+  const policyId = Cardano.AssetId.getPolicyId(assetId);
+  const assetInfo: Asset.AssetInfo = {
+    assetId,
+    fingerprint: Cardano.AssetFingerprint.fromParts(policyId, name),
+    name,
+    policyId,
+    quantity: 0n,
+    supply: amount
+  };
+
+  assetInfo.nftMetadata = getNftMetadata(name, policyId, tx, logger);
+
+  return assetInfo;
+};
+
+const getMintedAssetInfosFromTx = async (tx: Cardano.Tx, logger: Logger): Promise<Asset.AssetInfo[] | null> => {
+  const mints = tx.body.mint;
+
+  if (!mints) return null;
+
+  return [...mints.entries()]
+    .filter(([_, amount]) => amount > 0)
+    .map(([assetId, amount]) => createAssetInfo(assetId, amount, tx, logger));
+};
+
+const fetchAssetsFromProvider = async (
+  provider: AssetProvider,
+  assetIds: Cardano.AssetId[],
+  logger: Logger
+): Promise<Asset.AssetInfo[]> => {
+  const assetsFromProvider: Asset.AssetInfo[] = [];
+
+  // TODO: We need to fetch assets one by one because the provider will throw if any of the assets requests to the getAssets endpoint is not found.
+  // We want to fetch the ones we can and return a simplified AssetInfo for the ones we can't.
```

Comment:
Makes sense. Let's remove the 'TODO:' prefix of this comment

Reply:
done here https://github.com/input-output-hk/cardano-js-sdk/pull/1075/commits/5a6bfa60663fa702b84ed715740047fde129cc6f